### ISSUE TITLE
`kubectl merge` command - assert state for some fields, ignore others

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#37048]
Because of how patch is implemented, we can't easily manage things like configmaps through addon manager.  If a user edits the configmap and then addon manager applies the config again, the `data` field gets clobbered.  You just can't have 2 owners of an object like that.

What we need for this, and possibly others, seems to be an "overlay".  I want to assert a value for a handful of fields, and I want to ignore the rest.  The simplest case of this is a YAML that has an apiVersion, Kind, and metadata.name - this becomes an assertion of existence

**Comment by bgrant0607** [comment]
@thockin Do you have a specific use case? I don't think we should support this, so I'd like to find another way to solve the problem.

cc @pwittrock

**Comment by thockin** [comment]
We need(ed) this to make rollout of configmaps-for-addons easier.  I need
the configmap to exist for the kubelet to run the pod, but I can't install
the configmap from addon manager (kubectl apply) because it clobbers any
user-defined data (because data is a single field).  We worked around it by
reading the configmap directly instead of using kubelet.  We can work
around it other ways in the future - this seemed more elegant to me.

There are other cases where, thankfully, default values work for us, such
as deployments in addon manager.  The default replicas=1 means that we can
omit the replicas field entirely, and the new autoscaler will handle it.
This does not work, for example, with the vertical scaler (unless we
initially launch it as best effort and then add resources).

There were other examples, but I forget them right now.  @bowei or @mrhohn
might recall them.

**Comment by MrHohn** [comment]
I would like to elaborate the problem with addon manager a bit more.

Current addon manager replies on `kubectl apply` to fulfill its responsibility --- create and maintain resources with given fields. This pattern works well until we try to provide someone other than addon manager the ability to add/delete/modify non-given fields.

The situation we encountered is that when tried to use `kubectl edit` on addon objects, all added fields(non-given fields) were deleted by the `kubectl apply` from addon manager. Dug into the codes, found the root cause is `three way strategic merge patch`. Because `kubectl edit` will save the modified object into `kubectl.kubernetes.io/last-applied-configuration` annotation, which is latter used by the `kubectl apply`. Specifically in https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L1211, `kubectl apply` compares the object in this annotation with the given configuration we want to maintain. As the consequence, the added fields are put into `deletionsMap` and then being deleted.

As Tim said, the operation we really want is to assert state for given fields without clobbering others. In another word, to ensure given fields in a object without touching other fields.

**Comment by MrHohn** [comment]
Current horizontal autoscaler and vertical resizer work well with addon manager because they are both calling kubernetes client api directly, which will not has effect on `kubectl.kubernetes.io/last-applied-configuration` annotation.

**Comment by thockin** [comment]
I think the reason for this with ConfigMap is that `data` is a single
field, and maps get replaced, rather than merged?

**Comment by MrHohn** [comment]
Yes, if `data` field is defined in configuration, `kubectl apply` will replace the complete map instead of merging it.

But we could also omit the entire `data` field in configuration for ConfigMap. If above issue is resolved, `kubectl apply` will not reset this field anymore(after `kubectl edit` adds the `data` field). I think this is good enough in many cases.

**Comment by jbeda** [comment]
We should look for patterns that are similar to the way that distributions/package managers handle config files in `/etc`.  Usually a warning is given if user config is going to be clobbered.  In addition, there is a mechanism to drop a file into a `foo.d` directory.

Could we do something similar with ConfigMaps?  Perhaps enhance the ConfigMap volume plugin to work against a label and merge all of the ConfigMaps in to one directory? Provide a similar way to merge/dedup when using configmaps for env variables?

**Comment by bgrant0607** [comment]
I'd be on board with adding an option to create to ignore "already exists" errors, much like delete can ignore "doesn't exist" errors, but it's not clear to me that this is the best solution for the add-on manager. Perhaps we're missing some APIs, or missing some automation, or the components should read resources directly from the API, or we should create a "break glass" model that hands ownership entirely to the user. More generally, I'd love to eliminate the pattern of storing add-on manifests as local files on the master.

Users don't have this problem for their own applications, even off-the-shelf ones. They have the opportunity to do something prior to creation. Typically, they just fork the configuration, or can perform one-time actions. Tools such as Helm also support parameterization, as will the Service Broker. Additionally, most applications support their own override mechanisms, such as flags overriding env vars.

As we learned with borgcfg, if users can both override arbitrary fields but also depend on defaults, then the underlying template basically can't ever be modified without potentially breaking users. The same is true for Docker images -- modifying base layers requires throwing away and regenerating all layers above them.

We've also had projects that attempted to build config systems that supported arbitrary numbers of overlays. It was not a comprehensible model and was abandoned. In general, people settle for 2 layers, and that's sufficient if configuration is factored such that configuration of orthogonal knobs is composable.

Common patterns and scenarios are described here (shared with kubernetes-dev and several SIGs):
https://docs.google.com/document/d/1S3l2F40LCwFKg6WG0srR6056IiZJBwDmDvzHWRffTWk/edit#

For the fields that are commonly set by automation, such as replicas, cpu and memory requests/limits, and image, we should allow creation without those fields specified and add polymorphic subresources for imperative updates by automation components, similar to /scale.

**Comment by bgrant0607** [comment]
But if someone really wanted to do this, it would be trivial to parameterize the apply annotation, which would enable multiple appliers.

**Comment by bgrant0607** [comment]
@MrHohn If data is replaced rather than merged, that's just a bug.

cc @pwittrock
